sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
 from ...utils._typedefs cimport ITYPE_t, DTYPE_t
 from ...utils._typedefs import ITYPE, DTYPE
 
-cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_from_distance_matrix(
-    cnp.ndarray[cnp.double_t, ndim=2] distance_matrix
+# Numpy structured dtype representing a single ordered edge in Prim's algorithm
+MST_edge_dtype = np.dtype([
+    ("current_node", np.intp),
+    ("next_node", np.intp),
+    ("distance", np.float64),
+])
+
+# Packed shouldn't make a difference since they're all 8-byte quantities,
+# but it's included just to be safe.
+ctypedef packed struct MST_edge_t:
+    cnp.intp_t current_node
+    cnp.intp_t next_node
+    cnp.float64_t distance
+
```

Comment:
Also, with this struct (i.e. `MST_edge_t`), an Array of Structures (AoS) representation is used. An alternative representation using one array per fields, a Structure of Arrays (SoA) might be usable and used.

Is AoS more adapted than SoA for this implementation from a complexity (e.g. regarding algorithm time complexity) and technical (e.g. regarding caches' usages) perspective?

Note that I am am naively asking as I am not knowledgeable enough about HDBSCAN to know whether using SoA even makes sense.

Reply:
I believe the AoS approach is preferable because accesses are always done to all three fields, one struct at a time, e.g. https://github.com/scikit-learn/scikit-learn/blob/39e7d7e59dfdcf52488bd0d05dc7c43e67becef2/sklearn/cluster/_hdbscan/_linkage.pyx#L169-L171

Side note: Cython doesn't natively support a struct of buffers, so while this could be done with pointers, an SoA would introduce a fair bit of complexity in the form of memory management, no?

Reply:
Let's go with AoS then.

To answer your side note question: for SoA, you do not need to have a `struct`, but only several arrays of homogeneous elements (here 3 arrays for respectively `current_nodes`, `next_nodes` and `distances`). I think SoA (which is used in `PairwiseDistancesReductions`) is mainly convenient when you need to return independent arrays, but AoS is generally more adapted for algorithms.